crates/l1tx/src/filter/indexer.rs

```diff
@@ -0,0 +1,111 @@
+use bitcoin::{Block, Transaction};
+use strata_state::{
+    batch::SignedBatchCheckpoint,
+    tx::{DepositInfo, DepositRequestInfo},
+};
+
+use super::{
+    parse_checkpoint_envelopes, parse_da_blobs, parse_deposit_requests, parse_deposits,
+    TxFilterConfig,
+};
+use crate::messages::{DaEntry, L1BlockExtract, L1TxExtract, ProtocolTxEntry};
+
+pub trait TxIndexer {
+    // Do stuffs with `SignedBatchCheckpoint`.
+    fn visit_checkpoint(&mut self, _chkpt: SignedBatchCheckpoint) {}
+
+    // Do stuffs with `DepositInfo`.
+    fn visit_deposit(&mut self, _d: DepositInfo) {}
+
+    // Do stuffs with `DepositRequest`.
+    fn visit_deposit_request(&mut self, _d: DepositRequestInfo) {}
+
+    // Do stuffs with DA.
+    fn visit_da<'a>(&mut self, _d: impl Iterator<Item = &'a [u8]>) {}
+
+    // Collect data
+    fn collect(self) -> L1TxExtract;
+}
+
+pub trait BlockIndexer {
+    fn index_tx(&mut self, txidx: u32, tx: &Transaction, config: &TxFilterConfig);
+
+    fn index_block(mut self, block: &Block, config: &TxFilterConfig) -> Self
+    where
+        Self: Sized,
+    {
+        for (i, tx) in block.txdata.iter().enumerate() {
+            self.index_tx(i as u32, tx, config);
+        }
+        self
+    }
+
+    // Collect data
+    fn collect(self) -> L1BlockExtract;
+}
+
+#[derive(Clone, Debug)]
+pub struct OpIndexer<V: TxIndexer> {
+    visitor: V,
+    tx_entries: Vec<ProtocolTxEntry>,
+    dep_reqs: Vec<DepositRequestInfo>,
+    da_entries: Vec<DaEntry>,
+}
+
+impl<V: TxIndexer> OpIndexer<V> {
+    pub fn new(visitor: V) -> Self {
+        Self {
+            visitor,
+            tx_entries: Vec::new(),
+            dep_reqs: Vec::new(),
+            da_entries: Vec::new(),
+        }
+    }
+
+    pub fn tx_entries(&self) -> &[ProtocolTxEntry] {
+        &self.tx_entries
+    }
+
+    pub fn deposit_requests(&self) -> &[DepositRequestInfo] {
+        &self.dep_reqs
+    }
+
+    pub fn da_entries(&self) -> &[DaEntry] {
+        &self.da_entries
+    }
+}
+
+impl<V: TxIndexer + Clone> BlockIndexer for OpIndexer<V> {
+    fn index_tx(&mut self, txidx: u32, tx: &Transaction, config: &TxFilterConfig) {
```

Comment:
Once again, this it doesn't make any sense to take the config as an argument.  It only makes sense for it to be a field on the thing that makes the basic calls to the other utility fns that need it for extracting data from txs.

Reply:
The intent with this was to have the API explicitly imply that indexing should happen based on the `TxFilterConfig`. But as I think more about it now, it should be fine even if we bake that into the `OpIndexer`.